hamilton/plugins/h_mlflow.py

```diff
@@ -0,0 +1,295 @@
+            for node_name, value in value_sets.items():
+                self.client.log_param(self.run_id, key=node_name, value=value)
+
+        self.final_vars = final_vars
+
+    # TODO log DataLoaders as MLFlow datasets
+    def run_after_node_execution(
+        self,
+        *,
+        node_name: str,
+        node_return_type: Type,
+        node_tags: dict,
+        node_kwargs: dict,
+        result: Any,
+        **kwargs,
+    ):
+        """Log materializers and final vars as artifacts"""
+        # log DataSavers as artifacts
+        if node_tags.get("hamilton.data_saver") is True:
+            # don't log mlflow materializers as artifact since they already create models
+            # instead, use the Materializer metadata to add metadata to registered models
+            if node_tags["hamilton.data_saver.sink"] == "mlflow":
+                # skip if not registered model
+                if "registered_model" not in result.keys():
+                    return
+
+                # get the registered model name (param of MLFlowModelSaver)
+                model_name = result["registered_model"]["name"]
+                version = result["registered_model"]["version"]
+                materializer_node = self.graph[node_name]
+                # get the "materialized node" defining the model
+                materialized_node = self.graph[materializer_node.required_dependencies.pop()]
+                # add the materialized node docstring as description
+                # registered models have multiple versions
+                self.client.update_registered_model(model_name, materialized_node.documentation)
+                self.client.update_model_version(
+                    model_name, version, materialized_node.documentation
+                )
+
+                # add the materialized node @tag values as tags
+                for k, v in materialized_node.tags.items():
+                    # skip internal Hamilton tags
+                    if "hamilton." in k:
+                        continue
+                    self.client.set_registered_model_tag(model_name, key=k, value=v)
+                    self.client.set_model_version_tag(model_name, version, key=k, value=v)
+                # TODO automatically collect model input signature; maybe simpler from user code
+
+            # special case for matplotlib and plotly
+            # log materialized figure. Allows great degree of control over rendering format
+            # and also save interactive plotly visualization as HTML
+            elif node_tags["hamilton.data_saver.sink"] in ["plt", "plotly"]:
+                materializer_node = self.graph[node_name]
+                materialized_node = self.graph[materializer_node.required_dependencies.pop()]
+                figure = node_kwargs[materialized_node.name]
+
+                path = get_path_from_metadata(result)
+                if path:
+                    self.client.log_figure(self.run_id, figure, path)
```

Comment:
Hmm, this feels funky. Why are we logging both the figure and the path?

Reply:
This is the MLFlow API, not mine. You can read the line as 'write `figure` to `f"{self.run_id}/{path}"` '